ksqldb-engine/src/test/java/io/confluent/ksql/execution/ExpressionEvaluatorParityTest.java

```diff
@@ -0,0 +1,381 @@
+package io.confluent.ksql.execution;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.reset;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.engine.rewrite.ExpressionTreeRewriter;
+import io.confluent.ksql.engine.rewrite.ExpressionTreeRewriter.Context;
+import io.confluent.ksql.execution.codegen.CodeGenRunner;
+import io.confluent.ksql.execution.interpreter.InterpretedExpressionFactory;
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.expression.tree.QualifiedColumnReferenceExp;
+import io.confluent.ksql.execution.expression.tree.UnqualifiedColumnReferenceExp;
+import io.confluent.ksql.execution.expression.tree.VisitParentExpressionVisitor;
+import io.confluent.ksql.execution.transform.ExpressionEvaluator;
+import io.confluent.ksql.function.TestFunctionRegistry;
+import io.confluent.ksql.logging.processing.ProcessingLogger;
+import io.confluent.ksql.logging.processing.ProcessingLogger.ErrorMessage;
+import io.confluent.ksql.logging.processing.RecordProcessingError;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.name.ColumnName;
+import io.confluent.ksql.name.SourceName;
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlParserTestUtil;
+import io.confluent.ksql.util.MetaStoreFixture;
+import java.math.BigDecimal;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class ExpressionEvaluatorParityTest {
```

Comment:
I generally agree with this.  It would be nice to just take advantage of existing tests and not have to create a special way to write certain tests. The code is only set up to run pull queries with the new interpreter, so QTTs won't exercise it.  I agree that it would eventually be nice to use the interpreter everywhere if the performance is good for persistent queries (which my initial testing implies it should).  I just didn't want to go all in outside of pull queries just yet.  Also, any piece of functionality in the short term where there isn't parity will end up breaking.

My thinking would be to either run the RQTT with the new interpreter, or run RQTT with both.  That's not going to exercise the evaluation quite as much since, as you said, that's not their focus. I think even introducing a config to allow for interpreter use in persistent queries would be a non-trivial change since code gen is sprinkled throughout the codebase.  For that reason, I think it's worthwhile to maintain this `ExpressionEvaluatorParityTest` test until we do that change.

I'm open to more ideas, if you have any.

Reply:
Discussed a little offline but wanted to comment here. I think you've done an excellent job with the abstractions, so it shouldn't be too hard to make the persistent query code run with the interpreter (I don't think running the RQTTs would be that valuable, they don't really test expression evaluation that much).

I also wasn't thinking that we enable it for persistent queries in the near-term (we'd want to do a good chunk more testing, but I do think this should be the long term approach). Instead, I think we should have two configs - one for pull queries and one for persistent, the second config would only be used for testing at the moment to allow us to run QTTs using the interpreter.

This can be done in a separate PR, and perhaps timeboxed, but I'm pretty confident that we'd be able to get a good signal pretty quickly - even if we don't check it in to automatically run all the QTTs twice (having it be a manual switch might be good enough for the short-term).

Reply:
Until that point, I think keeping this test makes sense.